### ISSUE TITLE
"Combined" feeds aren't validated by CrossFeedDescriptorValidator 

### ISSUE THREAD

**Issue by barbeau** [CUTR-at-USF/gtfs-realtime-validator#85]
**Summary:** 

Historically, there have been three types of GTFS-rt feeds, each of which have been maintained in separate URLs:

1. TripUpdates
2. VehiclePositions
3. ServiceAlerts

However, there have been conversations on the Google Groups (e.g., https://groups.google.com/d/msg/gtfs-realtime/_HtNTGp5LxM/D07MFG_6hwQJ) saying that it is better for agencies to supply both TripUpdates and VehiclePosition entities within the same feed URL.  This allows for better coupling between the two types of objects that may cross-reference each other, and ensures that consumers have an atomic update for this information (vs. two different URLs that need to be cross-referenced).

The concepts of combined feeds didn't exist when this project was first created, so we should check to ensure that we can support combined feeds (in particular, TripUpdates + VehiclePosition entities in the same feed) in this validator.

Known issues with combined feeds:
* One known issue is that when we fixed the mixed messages issue (https://github.com/CUTR-at-USF/gtfs-realtime-validator/issues/256) via PR https://github.com/CUTR-at-USF/gtfs-realtime-validator/pull/270, if there is only a single GTFS-rt feed URL the `CrossFeedDescriptorValidator` won't be executed on that single GTFS-rt feed URL.  So, no cross feed errors would be detected.

**Steps to reproduce:** 

Pass in a combined feed into the validator ~(we may need to create one - I don't know of one in the wild off the top of my head)~.  

Combined Feed:
* GTFS - https://gtfsrt.api.translink.com.au/GTFS/SEQ_GTFS.zip
* GTFS-rt - https://gtfsrt.api.translink.com.au/Feed/SEQ

**Expected behavior:** 

Correctly validate the feed, including:
* If there is only a single GTFS-rt URL being monitored, but it's a combined feed, then `CrossFeedDescriptorValidator` should be run on the data.

**Observed behavior:** 

* If there is only a single GTFS-rt feed URL the `CrossFeedDescriptorValidator` won't be executed on that single GTFS-rt feed URL.  So, no cross feed errors would be detected.
* Maybe more problems???

**Comment by barbeau** [comment]
FYI - Official changes to GTFS-rt spec that allowed combined feeds are https://github.com/google/transit/commit/8be19ccbf1f5c6fd7c79519dd55378e8a653651b and https://github.com/google/transit/commit/c5b08b7da1f0ae2a3781b04dfc07f78edf80582c.

**Comment by barbeau** [comment]
I posted to the GTFS-realtime Google Group looking for examples of real-world combined feeds to test with here:
https://groups.google.com/forum/#!topic/gtfs-realtime/7zEiqvNxawQ

**Comment by barbeau** [comment]
Note that if we need to create our own feed to test, we can use https://github.com/CUTR-at-USF/gtfs-realtime-test-service.

**Comment by barbeau** [comment]
From City of Madison:

>Our real-time vendor has a "default" output directory, that may include this combined file you are describing - feel free to experiment as you see fit:
>* http://transitdata.cityofmadison.com/GTFS-RealTime/
>
>Traditional files:
>* http://transitdata.cityofmadison.com/GTFS/mmt_gtfs.zip
>* http://transitdata.cityofmadison.com/TripUpdate/
>* http://transitdata.cityofmadison.com/Vehicle/

**Comment by barbeau** [comment]
Looks like this is a combined feed with TripUpdates and VehiclePositions:
* GTFS - https://gtfsrt.api.translink.com.au/GTFS/SEQ_GTFS.zip
* GTFS-rt - https://gtfsrt.api.translink.com.au/Feed/SEQ

Feed excerpt (TripUpdate followed by VehiclePosition):

~~~
 {
      "id": "TU-9869636-QR 17_18-JUL_Mon-Monday-00-1R90",
      "trip_update": {
        "trip": {
          "4": [
            5
          ],
          "trip_id": "9869636-QR 17_18-JUL_Mon-Monday-00-1R90",
          "start_time": "23:39:00",
          "start_date": "20170731",
          "route_id": "SHBR-829"
        },
        "stop_time_update": [
          {
            "stop_sequence": 1,
            "departure": {
              "delay": 4,
              "time": 1501508344,
              "uncertainty": 30
            },
            "stop_id": "600425"
          },
          {
            "stop_sequence": 2,
            "arrival": {
              "delay": -46,
              "time": 1501508414,
              "uncertainty": 30
            },
            "departure": {
              "delay": 4,
              "time": 1501508464,
              "uncertainty": 30
            },
            "stop_id": "600423"
          },
          {
            "stop_sequence": 3,
            "arrival": {
              "delay": -90,
              "time": 1501508490,
              "uncertainty": 30
            },
            "departure": {
              "delay": -11,
              "time": 1501508569,
              "uncertainty": 30
            },
            "stop_id": "600421"
          },
          {
            "stop_sequence": 4,
            "arrival": {
              "delay": -82,
              "time": 1501508618,
              "uncertainty": 30
            },
            "departure": {
              "delay": 1,
              "time": 1501508701,
              "uncertainty": 30
            },
            "stop_id": "600419"
          },
          {
            "stop_sequence": 5,
            "arrival": {
              "delay": -93,
              "time": 1501508727,
              "uncertainty": 30
            },
            "departure": {
              "delay": 24,
              "time": 1501508844,
              "uncertainty": 30
            },
            "stop_id": "600417"
          },
          {
            "stop_sequence": 6,
            "arrival": {
              "delay": -29,
              "time": 1501508971,
              "uncertainty": 30
            },
            "departure": {
              "delay": 19,
              "time": 1501509019,
              "uncertainty": 30
            },
            "stop_id": "600415"
          },
          {
            "stop_sequence": 7,
            "arrival": {
              "delay": -37,
              "time": 1501509083,
              "uncertainty": 30
            },
            "departure": {
              "delay": 8,
              "time": 1501509128,
              "uncertainty": 30
            },
            "stop_id": "600413"
          },
          {
            "stop_sequence": 8,
            "arrival": {
              "delay": -46,
              "time": 1501509194,
              "uncertainty": 30
            },
            "departure": {
              "delay": 7,
              "time": 1501509247,
              "uncertainty": 30
            },
            "stop_id": "600411"
          },
          {
            "stop_sequence": 9,
            "arrival": {
              "delay": 103,
              "time": 1501509523,
              "uncertainty": 30
            },
            "departure": {
              "delay": 152,
              "time": 1501509572,
              "uncertainty": 30
            },
            "stop_id": "600407"
          },
          {
            "stop_sequence": 10,
            "arrival": {
              "delay": 138,
              "time": 1501509678,
              "uncertainty": 30
            },
            "departure": {
              "delay": 138,
              "time": 1501509678,
              "uncertainty": 30
            },
            "stop_id": "600403"
          },
          {
            "stop_sequence": 11,
            "arrival": {
              "delay": 87,
              "time": 1501509747,
              "uncertainty": 30
            },
            "departure": {
              "delay": 128,
              "time": 1501509788,
              "uncertainty": 30
            },
            "stop_id": "600397"
          },
          {
            "stop_sequence": 12,
            "arrival": {
              "delay": 127,
              "time": 1501509907,
              "uncertainty": 30
            },
            "departure": {
              "delay": 188,
              "time": 1501509968,
              "uncertainty": 30
            },
            "stop_id": "600395"
          },
          {
            "stop_sequence": 13,
            "arrival": {
              "delay": 120,
              "time": 1501510020,
              "uncertainty": 30
            },
            "departure": {
              "delay": 158,
              "time": 1501510058,
              "uncertainty": 30
            },
            "stop_id": "600391"
          },
          {
            "stop_sequence": 14,
            "arrival": {
              "delay": 111,
              "time": 1501510131,
              "uncertainty": 30
            },
            "departure": {
              "delay": 148,
              "time": 1501510168,
              "uncertainty": 30
            },
            "stop_id": "600365"
          },
          {
            "stop_sequence": 15,
            "arrival": {
              "delay": 73,
              "time": 1501510333,
              "uncertainty": 30
            },
            "departure": {
              "delay": 108,
              "time": 1501510368,
              "uncertainty": 30
            },
            "stop_id": "600005"
          },
          {
            "stop_sequence": 16,
            "arrival": {
              "delay": 108,
              "time": 1501510488,
              "uncertainty": 30
            },
            "departure": {
              "delay": 108,
              "time": 1501510488,
              "uncertainty": 30
            },
            "stop_id": "600014"
          },
          {
            "stop_sequence": 17,
            "arrival": {
              "delay": 70,
              "time": 1501510630,
              "uncertainty": 30
            },
            "departure": {
              "delay": 78,
              "time": 1501510698,
              "uncertainty": 30
            },
            "stop_id": "600017"
          },
          {
            "stop_sequence": 18,
            "arrival": {
              "delay": 348,
              "time": 1501511148,
              "uncertainty": 30
            },
            "stop_id": "600028"
          }
        ],
        "vehicle": {
          "id": "556990860_1R90"
        },
        "timestamp": 1501512019
      }
    },
    {
      "id": "VU-556990860_1572",
      "vehicle": {
        "trip": {
          "trip_id": "9869677-QR 17_18-JUL_Mon-Monday-00-1572",
          "route_id": "CAIP-829"
        },
        "position": {
          "latitude": -27.612925,
          "longitude": 152.76068
        },
        "timestamp": 1501511301,
        "stop_id": "600347",
        "vehicle": {
          "id": "556990860_1572",
          "label": "1572"
        }
      }
    },
~~~

**Comment by barbeau** [comment]
When we fixed the mixed messages issue (https://github.com/CUTR-at-USF/gtfs-realtime-validator/issues/256) via PR https://github.com/CUTR-at-USF/gtfs-realtime-validator/pull/270, now if there is only a single GTFS-rt feed URL the `CrossFeedDescriptorValidator` won't be executed on that single GTFS-rt feed URL.  So, no cross feed errors would be detected.  So, this is one known problem with combined feeds - I updated the initial post at the top of the issue with this information.
